tokio/src/util/linked_list.rs

```diff
@@ -4,95 +4,121 @@
+pub(crate) unsafe trait Link {
+    /// Handle to the list entry.
+    ///
+    /// This is usually a pointer-ish type.
+    type Handle;
+
+    /// Node type
+    type Target;
 
-/// A node which carries data of type `T` and is stored in an intrusive list.
+    /// Convert the handle to a raw pointer
+    ///
+    /// Consumes ownership of the handle.
+    fn to_raw(handle: Self::Handle) -> NonNull<Self::Target>;
+
+    /// Convert the raw pointer to a handle
+    fn from_raw(ptr: NonNull<Self::Target>) -> Self::Handle;
+
+    /// Return the pointers for a node
+    unsafe fn pointers(target: NonNull<Self::Target>) -> NonNull<Pointers<Self::Target>>;
+}
+
+/// Previous / next pointers
 #[derive(Debug)]
-pub(crate) struct Entry<T> {
+pub(crate) struct Pointers<T> {
     /// The previous node in the list. null if there is no previous node.
-    prev: Option<NonNull<Entry<T>>>,
+    prev: Option<NonNull<T>>,
 
     /// The next node in the list. null if there is no previous node.
-    next: Option<NonNull<Entry<T>>>,
-
-    /// The data which is associated to this list item
-    data: UnsafeCell<T>,
-
-    /// Prevents `Entry`s from being `Unpin`. They may never be moved, since
-    /// the list semantics require addresses to be stable.
-    _pin: PhantomPinned,
+    next: Option<NonNull<T>>,
 }
 
-unsafe impl<T: Send> Send for Entry<T> {}
-unsafe impl<T: Sync> Sync for Entry<T> {}
+unsafe impl<T: Send> Send for Pointers<T> {}
+unsafe impl<T: Sync> Sync for Pointers<T> {}
+
+// ===== impl LinkedList =====
 
-impl<T> LinkedList<T> {
+impl<T: Link> LinkedList<T> {
     /// Creates an empty linked list
-    pub(crate) fn new() -> Self {
+    pub(crate) fn new() -> LinkedList<T> {
         LinkedList {
             head: None,
             tail: None,
+            _p: PhantomData,
         }
     }
 
-    /// Adds an item to the back of the linked list.
-    ///
-    /// # Safety
-    ///
-    /// The function is only safe as long as valid pointers are stored inside
-    /// the linked list.
-    pub(crate) unsafe fn push_front(&mut self, entry: Pin<&mut Entry<T>>) {
-        let mut entry: NonNull<Entry<T>> = entry.get_unchecked_mut().into();
+    /// Adds an element first in the list.
+    pub(crate) fn push_front(&mut self, val: T::Handle) {
+        let ptr = T::to_raw(val);
 
-        entry.as_mut().next = self.head;
-        entry.as_mut().prev = None;
+        unsafe {
+            T::pointers(ptr).as_mut().next = self.head;
+            T::pointers(ptr).as_mut().prev = None;
 
-        if let Some(head) = &mut self.head {
-            head.as_mut().prev = Some(entry);
-        }
+            if let Some(head) = self.head {
+                T::pointers(head).as_mut().prev = Some(ptr);
+            }
 
-        self.head = Some(entry);
+            self.head = Some(ptr);
 
-        if self.tail.is_none() {
-            self.tail = Some(entry);
+            if self.tail.is_none() {
+                self.tail = Some(ptr);
+            }
         }
```

Comment:
nit: it seems like we could reduce the scope of the unsafe block here if we did
```rust
let pointers = unsafe { T::pointers(ptr).as_mut() };
pointers.next = self.head;
pointers.prev = None;

if let Some(head) = self.head {
    let head_pointers = unsafe { T::pointers(head).as_mut() };
    head_pointers.prev = Some(ptr);
}

// ...
```

Reply:
I considered this. I opted for the style above to avoid binding `&mut` references. Doing so increases the likelihood of accidental mut aliasing, which would be bad.